telcell/cell_identity.py

```diff
@@ -0,0 +1,295 @@
+import re
+from typing import Optional, Any
+
+RADIO_GSM = "GSM"
+RADIO_UMTS = "UMTS"
+RADIO_LTE = "LTE"
+RADIO_NR = "NR"
+
+CELL_IDENTITY_PATTERN = re.compile(
+    r'^((?P<radio>[a-zA-Z]+)/)?(?P<mcc>[0-9]+|\?)-(?P<mnc>[0-9]+|\?)'
+    + r'(-((?P<lac>[0-9]+|\?)-(?P<ci>[0-9]+|\?)|(?P<eci>[0-9]+|\?)))?$')
+
+
+class CellIdentity:
+    """
+    An antenna is identified by a Cell Global Identity (CGI) or an E-UTRAN Cell
+    Global Identity (eCGI), both 15 digit codes.
+
+    Resouces:
+        * https://www.cellmapper.net/
+    """
+
+    @staticmethod
+    def create(
+        radio: Optional[str] = None,
+        mcc: Optional[int] = None,
+        mnc: Optional[int] = None,
+        lac: Optional[int] = None,
+        ci: Optional[int] = None,
+        eci: Optional[int] = None,
+    ) -> "CellIdentity":
+        radio = radio.upper() if radio is not None else None
+
+        if radio == RADIO_GSM:
+            return GSMCell(mcc, mnc, lac, ci)
+        elif radio == RADIO_UMTS:
+            return UMTSCell(mcc, mnc, lac, ci)
+        elif radio == RADIO_LTE:
+            return LTECell(mcc, mnc, eci)
+        elif radio == RADIO_NR:
+            return NRCell(mcc, mnc, eci)
+        elif radio is not None:
+            raise ValueError(f"unsupported radio technology: {radio}")
+        elif eci is not None and (ci is None or lac is None):
+            # we have `eci` but not `lac`, `ci` --> guess LTE/NR
+            return EutranCellGlobalIdentity(mcc, mnc, eci)
+        elif ci is not None and lac is not None and eci is None:
+            # we have `lac` and `ci` but not `eci` --> guess GSM/UMTS
+            return CellGlobalIdentity(mcc, mnc, lac, ci)
+        elif eci is None and lac is None and ci is None:
+            return CellIdentity(mcc, mnc)  # guess it's a cell
+        else:
+            raise ValueError("either `ci` or `eci` should be provided, but not both")
+
+    @staticmethod
+    def parse(spec: str) -> "CellIdentity":
+        m = CELL_IDENTITY_PATTERN.match(spec)
+
+        def convert_value(key: str, value: str) -> Any:
+            if value is None or value == "?":
+                return None
+            elif key == "radio":
+                return value
+            else:
+                return int(value)
+
+        groups = {key: convert_value(key, value) for key, value in m.groupdict().items()}
+        return CellIdentity.create(**groups)
+
+    def __init__(self, mcc: Optional[int], mnc: Optional[int]):
+        """
+        General info on mobile network cell identifiers:
+
+            * [https://arimas.com/cgi-ecgi/]
+            * [http://www.techtrained.com/the-ultimate-cheat-sheet-for-lte-identifiers/]
+
+        Parameters:
+            mcc: Mobile Country Code (MCC, 3 digits)
+            mnc: Mobile Network Code (MNC, 2 digits)
+        """
+        assert mcc is None or isinstance(mcc, int), "mcc must be of type `int`"
+        assert mnc is None or isinstance(mnc, int), "mnc must be of type `int`"
```

Comment:
Other bits use an `if` statement. Considering that assertions outside of tests are frowned upon, maybe use an `if` here too?

Reply:
done, although so far nobody convinced me that using `assert` is bad practice